libraries/lib-stretching-sequence/ClipSegment.cpp

```diff
@@ -0,0 +1,116 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  ClipSegment.cpp
+
+  Matthieu Hodgkinson
+
+**********************************************************************/
+#include "ClipSegment.h"
+#include "ClipInterface.h"
+#include "SampleFormat.h"
+#include "StaffPadTimeAndPitch.h"
+
+#include <cassert>
+
+namespace
+{
+TimeAndPitchInterface::Parameters
+GetStretchingParameters(const ClipInterface& clip)
+{
+   TimeAndPitchInterface::Parameters params;
+   params.timeRatio = clip.GetStretchRatio();
+   return params;
+}
+
+sampleCount GetLastReadSample(
+   const ClipInterface& clip, double durationToDiscard,
+   PlaybackDirection direction)
+{
+   if (direction == PlaybackDirection::forward)
+      return sampleCount {
+         clip.GetRate() * durationToDiscard / clip.GetStretchRatio() + .5
+      };
+   else
+      return clip.GetPlaySamplesCount() - sampleCount {
+         clip.GetRate() * durationToDiscard / clip.GetStretchRatio() + .5
+      };
+}
+
+sampleCount
+GetTotalNumSamplesToProduce(const ClipInterface& clip, double durationToDiscard)
+{
+   return sampleCount { (clip.GetPlaySamplesCount().as_double() -
+                         durationToDiscard * clip.GetRate()) *
+                           clip.GetStretchRatio() +
+                        .5 };
+}
+} // namespace
+
+ClipSegment::ClipSegment(
+   const ClipInterface& clip, double durationToDiscard,
+   PlaybackDirection direction)
+    : mClip(clip)
+    , mLastReadSample(GetLastReadSample(clip, durationToDiscard, direction))
+    , mTotalNumSamplesToProduce(
+         GetTotalNumSamplesToProduce(clip, durationToDiscard))
+    , mPlaybackDirection(direction)
+    , mStretcher(std::make_unique<StaffPadTimeAndPitch>(
+         clip.GetWidth(), *this, GetStretchingParameters(clip)))
+{
+}
+
+size_t ClipSegment::GetFloats(float* const* buffers, size_t numSamples)
+{
+   const auto numSamplesToProduce =
+      std::min(
+         sampleCount { numSamples },
+         mTotalNumSamplesToProduce - mTotalNumSamplesProduced)
+         .as_size_t();
+   mStretcher->GetSamples(buffers, numSamplesToProduce);
+   mTotalNumSamplesProduced += numSamplesToProduce;
+   return numSamplesToProduce;
+}
+
+bool ClipSegment::Empty() const
+{
+   return mTotalNumSamplesProduced == mTotalNumSamplesToProduce;
```

Comment:
Maybe that should be `>=` just to be sure there is eventual emptiness even in the presence of an off-by-few error

Reply:
Similarly to above, `mTotalNumSamplesProduced` simply can't be larger than `mTotalNumSamplesToProduce`. I wouldn't want the reader to start doubting this.